open-craft/openedx-completion-aggregator
# [BB-2296] Add locking mechanism to batch operations

**Description:** This adds cache-based locking to batch operations. This way we can safely schedule management command with cron without special restrictions for preventing race conditions in auto-scaling groups.
It also fixes `reaggregate_course` management command by using [course](https://github.com/edx/edx-platform/blob/open-release/ironwood.master/common/djangoapps/student/models.py#L1025) instead of `course_key`.

**JIRA:** [BB-2296](https://tasks.opencraft.com/browse/BB-2296)

**Testing instructions:**
1. Install aggregator from this branch.
1. Uncomment `batch.py:74` (deleting the lock).
1. Run aggregator `/edx/bin/manage.edxapp lms run_aggregator_service` (if you have `StaleCompletions` on your devstack, you should run it twice).
1. Run the command from the previous point again. You should get warning `Aggregation is already running. Exiting.`.
1. (Optionally) Remove the lock with `from django.conf import settings; cache.delete(settings.COMPLETION_AGGREGATOR_AGGREGATION_LOCK)` and run aggregation service again.

**Reviewers:**

**Merge checklist:**

**Post merge:**
      finished.
